### PR TITLE
Allow cargo test to complete on OpenBSD 6.4/AMD64

### PR DESCRIPTION
**Pull request by pusateri**
disable some features that aren't included in OpenBSD.

### CODE REVIEW

File: test/sys/test_socket.rs [hunk 1]
```diff
@@ -117,6 +117,7 @@ pub fn test_socketpair() {
     assert_eq!(&buf[..], b"hello");
 }
 
+#[cfg(not(any(target_os = "openbsd")))]
```
Comment (asomers) [inline test/sys/test_socket.rs]: What's wrong with this test?

Comment (pusateri) [reply test/sys/test_socket.rs]: I didn't try to diagnose the problem. I was just trying to get tests to pass so I could work on IP6_PKTINFO, IP_RECVIF, and IP_RECV_DSTADDR
```
failures:

---- sys::test_socket::test_scm_rights stdout ----
thread 'sys::test_socket::test_scm_rights' panicked at 'slice index starts at 24 but ends at 20', libcore/slice/mod.rs:1977:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Comment (pusateri) [reply test/sys/test_socket.rs]: This could be #999. If there's an alignment problem, off by 4 could happen.

Comment (pusateri) [reply test/sys/test_socket.rs]: Here's a stack trace with RUST_BACKTRACE=1
```
failures:

---- sys::test_socket::test_scm_rights stdout ----
thread 'sys::test_socket::test_scm_rights' panicked at 'slice index starts at 24 but ends at 20', libcore/slice/mod.rs:1977:5
stack backtrace:
   0: __register_frame_info
   1: __register_frame_info
   2: __register_frame_info
   3: __register_frame_info
   4: __register_frame_info
   5: __register_frame_info
   6: __register_frame_info
   7: __register_frame_info
   8: __register_frame_info
   9: __register_frame_info
  10: __register_frame_info
  11: __register_frame_info
  12: __register_frame_info
  13: __register_frame_info
  14: __register_frame_info
  15: __register_frame_info
  16: __register_frame_info
  17: __register_frame_info
  18: __register_frame_info
  19: __register_frame_info
  20: __register_frame_info
  21: __register_frame_info
  22: __register_frame_info
  23: __register_frame_info
  24: pthread_create


failures:
    sys::test_socket::test_scm_rights

test result: FAILED. 87 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

Comment (pusateri) [reply test/sys/test_socket.rs]: When running the test as standalone code in a main(), I got a different stack trace:
```
% RUST_BACKTRACE=1 cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 4.53s
     Running `target/debug/scm_rights`
thread 'main' panicked at 'slice index starts at 24 but ends at 20', libcore/slice/mod.rs:1977:5
stack backtrace:
   0: __register_frame_info
   1: __register_frame_info
   2: __register_frame_info
   3: __register_frame_info
   4: __register_frame_info
   5: __register_frame_info
   6: __register_frame_info
   7: __register_frame_info
   8: __register_frame_info
   9: __register_frame_info
  10: __register_frame_info
  11: __register_frame_info
  12: __register_frame_info
  13: __register_frame_info
  14: __register_frame_info
  15: __register_frame_info
  16: __register_frame_info
  17: __register_frame_info
  18: __register_frame_info
  19: __register_frame_info
  20: <unknown>
```

Comment (pusateri) [reply test/sys/test_socket.rs]: stepping through the debugger, it crashes looping through the cmsgs:
```
(gdb) list
1       extern crate nix;
2       use std::os::unix::io::RawFd;
3       
4       pub fn main() {
5           use nix::sys::uio::IoVec;
6           use nix::unistd::{pipe, read, write, close};
7           use nix::sys::socket::{socketpair, sendmsg, recvmsg,
8                                  AddressFamily, SockType, SockFlag,
9                                  ControlMessage, CmsgSpace, MsgFlags};
10      
(gdb) 
11          let (fd1, fd2) = socketpair(AddressFamily::Unix, SockType::Stream, None, SockFlag::empty())
12                           .unwrap();
13          let (r, w) = pipe().unwrap();
14          let mut received_r: Option<RawFd> = None;
15      
16          {
17              let iov = [IoVec::from_slice(b"hello")];
18              let fds = [r];
19              let cmsg = ControlMessage::ScmRights(&fds);
20              assert_eq!(sendmsg(fd1, &iov, &[cmsg], MsgFlags::empty(), None).unwrap(), 5);
(gdb) break 11
Breakpoint 1 at 0x15ea4: file src/main.rs, line 11.
(gdb) run
Starting program: /home/pusateri/Projects/scm_rights/target/debug/scm_rights 

Breakpoint 1, scm_rights::main::h34fb823cda0cd73f () at src/main.rs:11
11          let (fd1, fd2) = socketpair(AddressFamily::Unix, SockType::Stream, None, SockFlag::empty())
(gdb) n
13          let (r, w) = pipe().unwrap();
(gdb) n
14          let mut received_r: Option<RawFd> = None;
(gdb) n
17              let iov = [IoVec::from_slice(b"hello")];
(gdb) 
warning: (Internal error: pc 0x833443cd400 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd400 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd400 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd400 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd400 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd400 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd400 in read in psymtab, but not in symtab.)

18              let fds = [r];
(gdb) 
19              let cmsg = ControlMessage::ScmRights(&fds);
(gdb) 
20              assert_eq!(sendmsg(fd1, &iov, &[cmsg], MsgFlags::empty(), None).unwrap(), 5);
(gdb) 
21              close(r).unwrap();
(gdb) 
22              close(fd1).unwrap();
(gdb) 
26              let mut buf = [0u8; 5];
(gdb) 
27              let iov = [IoVec::from_mut_slice(&mut buf[..])];
(gdb) 
warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x833443cd470 in read in psymtab, but not in symtab.)

28              let mut cmsgspace: CmsgSpace<[RawFd; 1]> = CmsgSpace::new();
(gdb) 
29              let msg = recvmsg(fd2, &iov, Some(&mut cmsgspace), MsgFlags::empty()).unwrap();
(gdb) 
31              for cmsg in msg.cmsgs() {
(gdb) 
thread 'main' panicked at 'slice index starts at 24 but ends at 20', libcore/slice/mod.rs:1977:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
[Inferior 1 (process 45666) exited with code 0145]
(gdb) 
```

Comment (pusateri) [reply test/sys/test_socket.rs]: More debug info:
```
Reading symbols from target/debug/scm_rights...done.
(gdb) list 31
26              let mut buf = [0u8; 5];
27              let iov = [IoVec::from_mut_slice(&mut buf[..])];
28              let mut cmsgspace: CmsgSpace<[RawFd; 1]> = CmsgSpace::new();
29              let msg = recvmsg(fd2, &iov, Some(&mut cmsgspace), MsgFlags::empty()).unwrap();
30      
31              for cmsg in msg.cmsgs() {
32                  if let ControlMessage::ScmRights(fd) = cmsg {
33                      assert_eq!(received_r, None);
34                      assert_eq!(fd.len(), 1);
35                      received_r = Some(fd[0]);
(gdb) break 31
Breakpoint 1 at 0x163ad: /home/pusateri/Projects/scm_rights/src/main.rs:31. (3 locations)
(gdb) run
Starting program: /home/pusateri/Projects/scm_rights/target/debug/scm_rights 

Breakpoint 1, scm_rights::main::h34fb823cda0cd73f () at src/main.rs:31
31              for cmsg in msg.cmsgs() {
(gdb) print msg
$1 = RecvMsg = {bytes = 5, cmsg_buffer = &[u8](len: 20) = {20, 0, 0, 0, 255, 
    255, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, Python Exception <class 'gdb.error'> Cannot convert value to int.: 
0}, 
  address = <error reading variable>, flags = MsgFlags = {bits = 0}}
(gdb) 
```

Comment (asomers) [reply test/sys/test_socket.rs]: I don't think #999 is related.  If anything, #999 should cause bus errors on platforms like mips that don't allow unaligned loads.  This error looks more like we got some alignment stuff wrong in #648.  When you encounter the bug, what is the address of the `struct cmsghdr`?  If it's 4-byte aligned, then try putting the `CmsgSpace` in a `Box`.  That should increase its alignment to 16 bytes.  If that fixes the test, then we know where the problem lies.

Comment (pusateri) [reply test/sys/test_socket.rs]: Looks 16 byte aligned to me. On the sendmsg side:
```
(gdb) print cmsg
Python Exception <class 'gdb.error'> That operation is not available on integers of more than 8 bytes.: 
Python Exception <class 'gdb.error'> That operation is not available on integers of more than 8 bytes.: 
$1 = nix::sys::socket::ControlMessage::ScmRights(&[i32](len: 1) = {5})
(gdb) print &cmsg
$2 = (nix::sys::socket::ControlMessage *) 0x7f7ffffea920
```
On the recvmsg side:
```
28              let mut cmsgspace: CmsgSpace<[RawFd; 1]> = CmsgSpace::new();
(gdb) n
29              let msg = recvmsg(fd2, &iov, Some(&mut cmsgspace), MsgFlags::empty()).unwrap();
(gdb) print &cmsgspace
$3 = (nix::sys::socket::CmsgSpace<[i32; 1]> *) 0x7f7ffffeaa60
```

Comment (asomers) [reply test/sys/test_socket.rs]: Well, there goes that theory.  Maybe `cmsg_align` doesn't work correctly on OpenBSD?  If you could reimplement the test in C, that might reveal where any alignment errors lie.

Comment (pusateri) [reply test/sys/test_socket.rs]: It looks like the unwrapped RecvMsg from recvmsg() is not 16 byte aligned:
```
Breakpoint 1, scm_rights::main::h34fb823cda0cd73f () at src/main.rs:34
34                  for cmsg in msg.cmsgs() {
(gdb) print msg
$1 = RecvMsg = {bytes = 5, cmsg_buffer = &[u8](len: 20) = {20, 0, 0, 0, 255, 
    255, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, Python Exception <class 'gdb.error'> Cannot convert value to int.: 
0}, 
  address = <error reading variable>, flags = MsgFlags = {bits = 0}}
(gdb) print &msg
$2 = (nix::sys::socket::RecvMsg *) 0x7f7fffff2838
```

Comment (asomers) [reply test/sys/test_socket.rs]: So I can fix this particular error by setting `align_of_cmsg_data = u32` on OpenBSD.  But I don't have any theoretical reason for doing that, and it breaks the ScmTimestamp test.  I think the best thing to do is to rewrite all of the cmsg code in terms of `CMSG_DATA` and friends, which are provided by libc.

**Comment by asomers** [comment]
Closing as a duplicate of #1000 .

**Pull request closed without merging**